flow/Trace.h

```diff
@@ -133,9 +134,196 @@ class TraceBatch {
+	}
+}
+
+// forward declare format from flow.h as we
+// can't include flow.h here
+std::string format(const char* form, ...);
+
+template<class T>
+struct Traceable : std::false_type {};
+
+#define FORMAT_TRACEABLE(type, fmt)					\
+	template<>										\
+	struct Traceable<type> : std::true_type {		\
+		static std::string toString(type value) {	\
+			return format(fmt, value);				\
+		}											\
+	}
+
+FORMAT_TRACEABLE(bool, "%d");
+FORMAT_TRACEABLE(signed char, "%d");
+FORMAT_TRACEABLE(unsigned char, "%d");
+FORMAT_TRACEABLE(short, "%d");
+FORMAT_TRACEABLE(unsigned short, "%d");
+FORMAT_TRACEABLE(int, "%d");
+FORMAT_TRACEABLE(unsigned, "%u");
+FORMAT_TRACEABLE(long int, "%ld");
+FORMAT_TRACEABLE(unsigned long int, "%lu");
+FORMAT_TRACEABLE(long long int, "%lld");
+FORMAT_TRACEABLE(unsigned long long int, "%llu");
+FORMAT_TRACEABLE(double, "%g");
+FORMAT_TRACEABLE(volatile long, "%ld");
+FORMAT_TRACEABLE(volatile unsigned long, "%lu");
+FORMAT_TRACEABLE(volatile long long, "%lld");
+FORMAT_TRACEABLE(volatile unsigned long long, "%llu");
+FORMAT_TRACEABLE(volatile double, "%g");
+
+
+template<>
+struct Traceable<UID> : std::true_type {
+	static std::string toString(const UID& value) {
+		return format("%016llx", value.first());
+	}
+};
+
+template<>
+struct Traceable<const char*> : std::true_type {
+	static bool isPrintable(char c) { return c > 32 && c < 127; }
+	static std::string toString(const char* value) {
+		// if all characters are printable ascii, we simply return the string
+		int nonPrintables = 0;
+		int numBackslashes = 0;
+		auto val = value;
+		int size = 0;
+		while (auto c = *(val++)) {
+			++size;
+			if (!Traceable<const char*>::isPrintable(c)) {
+				++nonPrintables;
+			} else if (c == '\\') {
+				++numBackslashes;
+			}
+		}
+		if (nonPrintables == 0 && numBackslashes == 0) {
+			return std::string(value);
+		}
+		std::string result;
+		result.reserve(size - nonPrintables + (nonPrintables * 4) + numBackslashes);
+		while (auto c = *(val++)) {
+			if (Traceable<const char*>::isPrintable(c)) {
+				result.push_back(c);
+			} else if (c == '\\') {
+				result.push_back('\\');
+				result.push_back('\\');
+			} else {
+				result.push_back('\\');
+				result.push_back('x');
+				result.push_back(base16Char((c / 16) % 16));
+				result.push_back(base16Char(c % 16));
+			}
+		}
+		return result;
+	}
+};
+
+template<size_t S>
+struct Traceable<char[S]> : std::true_type {
+	static std::string toString(const char* value) {
+		return Traceable<const char*>::toString(value);
+	}
+};
+
+template<>
+struct Traceable<char*> : std::true_type {
+	static std::string toString(const char* value) {
+		return Traceable<const char*>::toString(value);
+	}
+};
+
+template<>
+struct Traceable<std::string> : std::true_type {
+	static bool isPrintable(char c) { return c >= 32 && c < 127; }
+	template<class Str>
+	static std::string toString(Str&& value) {
+		// if all characters are printable ascii, we simply return the string
```

Comment:
I see we have multiple implementations of the printable logic. Any way these can be consolidated?

Reply:
I consolidated this into a template. Sadly, this got quite complex (new commits will follow shortly). Reason for that is that I want to be very careful to not make unnecessary copies.

It still has value imho, as the find-replace logic is now in one class.